malsim/mal_simulator.py

```diff
@@ -411,76 +412,76 @@ def _defender_step(
             if node.is_available_defense():
                 node.defense_status = 1.0
                 node.is_viable = False
-                attack_steps_made_unviable += \
+                attack_steps_made_unviable |= \
                     apriori.propagate_viability_from_unviable_node(node)
                 agent.reward -= node.extras.get("reward", 0)
-                enabled_defenses.append(node)
+                enabled_defenses.add(node)
                 logger.info(
                     'Defender agent "%s" enabled "%s"(%d).',
                     agent.name, node.full_name, node.id
                 )
 
-
         for node in enabled_defenses:
             # Remove enabled defenses from defender action surfaces
             for defender_agent in self._get_defender_agents():
-                try:
-                    defender_agent.action_surface.remove(node)
-                except ValueError:
-                    pass
+                defender_agent.action_surface.discard(node)
 
         for attack_step in attack_steps_made_unviable:
             # Remove unviable attack steps from attacker action surfaces
             for attacker_agent in self._get_attacker_agents():
-                try:
-                    attacker_agent.action_surface.remove(attack_step)
-                except ValueError:
-                    pass
+                attacker_agent.action_surface.discard(attack_step)
 
         if self.sim_settings.uncompromise_untraversable_steps:
             self._disable_attack_steps(attack_steps_made_unviable)
 
         return enabled_defenses, attack_steps_made_unviable
 
     def step(
-            self, actions: dict[str, list[AttackGraphNode]]
-        ) -> tuple[list[AttackGraphNode], list[AttackGraphNode]]:
+        # TODO: Do we care about order of requested actions to have a list?
```

Comment:
Yes, we do, if we allow an agent to take multiple actions at once they need to happen in order in case some of  the actions earlier in the order make possible actions later in the list.

Reply:
This may complicate things because the input to the step methods needs to be a list then, and these methods will return a set. Then, the agent needs to consume a set and return a list. Maybe it is simple to change that but I'll have to fix the code.

Reply:
Actually, I did not change the signatures of the args in the step methods :phew:, they are still lists.